sql/sql_connect.cc

```diff
@@ -670,6 +670,24 @@ void reset_mqh(THD *thd, LEX_USER *lu, bool get_them = false) {
 
 bool thd_init_client_charset(THD *thd, uint cs_number) {
   CHARSET_INFO *cs;
+
+  // if the 8.0 client sets 'MYSQL_SET_CHARSET_NAME' option to 'utf8mb4' or
+  // leaves it empty, basically meaning the same, this function will be called
+  // with 'cs_number' equal to 255 (meaning 'utf8mb4_0900_ai_ci')
+
+  // at the same time, if 'default_collation_for_utf8mb4' is set to something
+  // other than default 'utf8mb4' collation ('utf8mb4_0900_ai_ci', number 255),
+  // we need to fix 'cs_number' here by setting it to the corresponding number
+  // of 'default_collation_for_utf8mb4' (currently only 'utf8mb4_general_ci',
+  // number 45, is supported)
+  const auto *primary_utf8mb4_collation =
```

Comment:
Yes, @dlenev I had exactly the same doubts and was going back and forth with this. From one side I did not want to hardcode `primary_utf8mb4_collation` to be `my_charset_utf8mb4_0900_ai_ci` (because who knows may be in the next version the default collation will change again). On the other hand, I totally agree that establishing the connection is a critical path and we should not add any unnecessary cycles here.
Anyway, if this caught your attention as well, then it is probably more serious than I thought.
Let's wait for the final feedback from the customer and I will add the changes you suggested into the final patch.

Reply:
@dlenev I reworked the critical paths code with simplified versions
```c++
  if (thd->variables.default_collation_for_utf8mb4 !=
      &my_charset_utf8mb4_0900_ai_ci) {
    if (client_cs == &my_charset_utf8mb4_0900_ai_ci) {
      client_cs = thd->variables.default_collation_for_utf8mb4;
    }
  }
```
that does not involve charset by name lookup.